File: CS/ActionAttributeExample.Module/BusinessObjects/Task.cs
using DevExpress.ExpressApp.Model;
using System;
using System.ComponentModel;

using DevExpress.Xpo;
using DevExpress.Data.Filtering;

using DevExpress.ExpressApp;
using DevExpress.ExpressApp.DC;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;

namespace ActionAttributeExample.Module.BusinessObjects {
    [DefaultClassOptions, ImageName("BO_Task"), DefaultProperty("Subject")]
    public class Task : BaseObject {
        public Task(Session session) : base(session) { }
        private string subject;
        public string Subject {
            get { return subject; }
            set { SetPropertyValue("Subject", ref subject, value); }
        }
        private DateTime? deadline;
        public DateTime? Deadline {
            get { return deadline; }
            set { SetPropertyValue("Deadline", ref deadline, value); }
        }
        private bool isCompleted;
        public bool IsCompleted {
            get { return isCompleted; }
            set { SetPropertyValue("IsCompleted", ref isCompleted, value); }
        }
        private string comments;
        [Size(SizeAttribute.Unlimited), ModelDefault("AllowEdit", "False")]
        public string Comments {
            get { return comments; }
            set { SetPropertyValue("Comments", ref comments, value); }
        }
        [Action(Caption="Complete", TargetObjectsCriteria = "Not [IsCompleted]")]
        public void Complete() {
            IsCompleted = true;
        }
        [Action(Caption = "Postpone",
            TargetObjectsCriteria = "[Deadline] Is Not Null And Not [IsCompleted]")]
        public void Postpone(PostponeParametersObject parameters) {
            if (Deadline.HasValue && !IsCompleted && (parameters.PostponeForDays > 0)) {
                Deadline += TimeSpan.FromDays(parameters.PostponeForDays);
                Comments += String.Format("Postponed for {0} days, new deadline is {1:d}\r\n{2}\r\n",
                    parameters.PostponeForDays, Deadline, parameters.Comment);
            }
        }
    }
    [DomainComponent]
    [ModelDefault("Caption", "Postpone Parameters")]
    public class PostponeParametersObject {
        public PostponeParametersObject() { PostponeForDays = 1; }
        public uint PostponeForDays { get; set; }
        [Size(SizeAttribute.Unlimited)]
        public string Comment { get; set; }
    }
}

File: VB/ActionAttributeExample.Module/BusinessObjects/Task.vb
Imports DevExpress.ExpressApp.Model
Imports System
Imports System.ComponentModel

Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering

Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.DC
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl

Namespace ActionAttributeExample.Module.BusinessObjects
	<DefaultClassOptions, ImageName("BO_Task"), DefaultProperty("Subject")>
	Public Class Task
		Inherits BaseObject

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
'INSTANT VB NOTE: The field subject was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private subject_Conflict As String
		Public Property Subject() As String
			Get
				Return subject_Conflict
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Subject", subject_Conflict, value)
			End Set
		End Property
'INSTANT VB NOTE: The field deadline was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private deadline_Conflict? As DateTime
		Public Property Deadline() As DateTime?
			Get
				Return deadline_Conflict
			End Get
			Set(ByVal value? As DateTime)
				SetPropertyValue("Deadline", deadline_Conflict, value)
			End Set
		End Property
'INSTANT VB NOTE: The field isCompleted was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private isCompleted_Conflict As Boolean
		Public Property IsCompleted() As Boolean
			Get
				Return isCompleted_Conflict
			End Get
			Set(ByVal value As Boolean)
				SetPropertyValue("IsCompleted", isCompleted_Conflict, value)
			End Set
		End Property
'INSTANT VB NOTE: The field comments was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private comments_Conflict As String
		<Size(SizeAttribute.Unlimited), ModelDefault("AllowEdit", "False")>
		Public Property Comments() As String
			Get
				Return comments_Conflict
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Comments", comments_Conflict, value)
			End Set
		End Property
		<Action(Caption:="Complete", TargetObjectsCriteria := "Not [IsCompleted]")>
		Public Sub Complete()
			IsCompleted = True
		End Sub
		<Action(Caption := "Postpone", TargetObjectsCriteria := "[Deadline] Is Not Null And Not [IsCompleted]")>
		Public Sub Postpone(ByVal parameters As PostponeParametersObject)
			If Deadline.HasValue AndAlso Not IsCompleted AndAlso (parameters.PostponeForDays > 0) Then
				Deadline += TimeSpan.FromDays(parameters.PostponeForDays)
				Comments &= String.Format("Postponed for {0} days, new deadline is {1:d}" & vbCrLf & "{2}" & vbCrLf, parameters.PostponeForDays, Deadline, parameters.Comment)
			End If
		End Sub
	End Class
	<DomainComponent>
	<ModelDefault("Caption", "Postpone Parameters")>
	Public Class PostponeParametersObject
		Public Sub New()
			PostponeForDays = 1
		End Sub
		Public Property PostponeForDays() As UInteger
		<Size(SizeAttribute.Unlimited)>
		Public Property Comment() As String
	End Class
End Namespace
